src/host/readDataCooked.cpp

```diff
@@ -436,633 +196,1027 @@ bool COOKED_READ_DATA::AtEol() const noexcept
-                }
-                else
-                {
-                    _bytesRead = 0;
-                }
-                break;
-            }
-        }
-        else
+void COOKED_READ_DATA::_popupHandleCopyToCharInput(Popup& /*popup*/, const wchar_t wch, const uint16_t vkey, const DWORD /*modifiers*/)
+{
+    if (vkey == VK_ESCAPE)
+    {
+        _popupDone();
+        return;
+    }
+
+    if (wch)
+    {
+        const auto cmd = _history->GetLastCommand();
+        const auto idx = cmd.find(wch, _bufferCursor);
+
+        if (idx != decltype(cmd)::npos)
         {
-            if (ProcessInput(wch, keyState, Status))
-            {
-                auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-                gci.Flags |= CONSOLE_IGNORE_NEXT_KEYUP;
-                break;
-            }
+            // When we enter this if condition it's guaranteed that _bufferCursor must be
+            // smaller than idx, which in turn implies that it's smaller than cmd.size().
+            // As such, calculating length is safe and str.size() == length.
+            const auto count = idx - _bufferCursor;
+            _buffer.replace(_bufferCursor, count, cmd, _bufferCursor, count);
+            _bufferCursor += count;
+            _markAsDirty();
         }
+
+        _popupDone();
     }
-    return Status;
 }
 
-// Routine Description:
-// - handles any tasks that need to be completed after the read input loop finishes
-// Arguments:
-// - isUnicode - Treat as UCS-2 unicode or use Input CP to convert when done.
-// - numBytes - On in, the number of bytes available in the client
-// buffer. On out, the number of bytes consumed in the client buffer.
-// - controlKeyState - For some types of reads, this is the modifier key state with the last button press.
-// Return Value:
-// - Status code that indicates success, out of memory, etc.
-[[nodiscard]] NTSTATUS COOKED_READ_DATA::_handlePostCharInputLoop(const bool isUnicode, size_t& numBytes, ULONG& controlKeyState) noexcept
+void COOKED_READ_DATA::_popupHandleCopyFromCharInput(Popup& /*popup*/, const wchar_t wch, const uint16_t vkey, const DWORD /*modifiers*/)
 {
-    std::span writer{ _userBuffer, _userBufferSize };
-    std::wstring_view input{ _backupLimit, _bytesRead / sizeof(wchar_t) };
-    DWORD LineCount = 1;
+    if (vkey == VK_ESCAPE)
+    {
+        _popupDone();
+        return;
+    }
 
-    if (_echoInput)
+    if (wch)
     {
-        const auto idx = input.find(UNICODE_CARRIAGERETURN);
-        if (idx != decltype(input)::npos)
-        {
-            if (_commandHistory)
-            {
-                auto& gci = ServiceLocator::LocateGlobals().getConsoleInformation();
-                LOG_IF_FAILED(_commandHistory->Add({ _backupLimit, idx }, WI_IsFlagSet(gci.Flags, CONSOLE_HISTORY_NODUP)));
-            }
+        const auto idx = _buffer.find(wch, _bufferCursor);
+        _buffer.erase(_bufferCursor, std::min(idx, _buffer.size()) - _bufferCursor);

```

Comment:
This might want to retain some commentary. The original file had some comments that at least made it easier to glance at it:

```
        // char not found, delete everything to the right of the cursor

        // char was found, delete everything between the cursor and it

```